# Feedback from previous weeks and other and hints

1. Be careful about your environment remembering variables. Make sure your code works in a new _clean_ environment. In Colab: `Runtime`->`restart Runtime`, in Anaconda's Jupyter: `Kernel`->`Restart`.
2. Graphs without labels (or units when appropriate) are not worth any point.
3. Do put in sufficient explanatory comments in your code.
4. Functions are very important. Do look up the video on the Safari O'Reilly ressource if you are still not clear on them !

For this week you can use these imports at the start of your programs:

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

We will use a new module `uncertainties`, which is **not** standard in the colab environment. You will have to first run:

In [ ]:
! pip install -q uncertainties

To install the module before you can import it:

In [ ]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions

# Introduction
In the practical classes PX2133/PX2233 and PX2338 (Obs tech), as well as your year 3/4 project, a lot of emphasis is placed on the determination and mathematical handling of errors.
The uncertainties module allows us to deal very easily with [error propagation](https://en.wikipedia.org/wiki/Propagation_of_uncertainty). For this sheet you should remind yourself about error bars in measurements and about propagation of uncertainties. Take an example from your lab handbook:

**Example 1**: If the length of a rectangle is $1.24\pm0.02 m$ and its breadth is $0.61\pm0.01 m$, what is its area and the error in the area? The following code snippet solves this problem in a few lines.

In [ ]:
L = uc.ufloat(1.24, 0.02)
W = uc.ufloat(0.61, 0.01)
print ('Area is:', L*W, 'm^2') # Do remember to add the units when printing!

Area is: 0.756+/-0.017 m^2


**Note**: For the area itself, it's fairly straightforward:

In [ ]:
1.24*0.61

0.7564

However, for the error bar on this number:

In [ ]:
0.02*0.01

0.0002

does not work. Instead, the [error progation formula](https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Example_formulae) gives:

In [ ]:
np.abs(1.24*0.61)*np.sqrt((0.02/1.24)**2+(0.01/0.61)**2)

0.01739540169125163

So the area is $0.756\pm0.017 m^2$. `uncertainties` obviously saves a lot of work, even for such a simple case. You can also take a look at the web site uncertainties hosted at https://pythonhosted.org/uncertainties/user_guide.html. In particular, [this section](https://pythonhosted.org/uncertainties/user_guide.html#access-to-the-uncertainty-and-to-the-nominal-value) shows some of the properties of a `ufloat` you can access directly.

**Example 2**: A reference object is $10.0\pm0.0001 m$ long, and makes a viewing angle of $0.62\pm0.02 rad$. How far is it?

In [ ]:
L = uc.ufloat(10.0, 0.0001)
theta = uc.ufloat(0.62,0.02)

Distance = (L/2)/um.tan(theta/2)

print ('Distance is:', Distance.nominal_value, 'm, with an error of:', Distance.std_dev)

Distance is: 15.609024890896208 m, with an error of: 0.537283338762715


Note the need to use "umath" functions (like `um.tan()` instead of `np.tan()`), and how to get the nominal value and the standard deviation of the uncertainties objects. To get nicer looking output, such as controlling the number of significant digits printed, you can use the information about formatting at https://docs.python.org/3/tutorial/inputoutput.html. In the exercises below you need to print the values to the screen. (Don’t forget units.)

# Exercises
This must be marked before you leave the lab. Mark weighting is in brackets.
**Save your work to GitHub after having run all cells with `Runtime` -> `Restart and run all`. And do not change the notebook's filename.** Do add comments to your code, you'll lose points if your code is hard to understand. Graphs without labels (or units when appropriate) are not worth any point.

## Exercise 0
[0] With some approximations, we have measured the mass of the following black-holes:
```
"35.6+/-3.9","30.6+/-3.7","63.1+/-3.2","23.2+/-9.8","13.6+/-4.5","35.7+/-6.8","13.7+/-6.0","7.7+/-2.4","20.5+/-4.0"
```
Compute for each (with error-bars) their lifetime due to Hawking radiation:
$$
t = \left(\frac{M}{M_{\odot}}\right)^3\,\times\,2.097\,\times\,10^{67} yr
$$

(this exercise is for demonstration purposes and won't be marked)

## Exercise 1
[2] An object is measured to travel a distance $x = 5.1 \pm 0.4 m$ during a time of $t = 0.4 \pm 0.1 s$. What is the average velocity and the error in the average velocity?

In [1]:
# Given values
x = 5.1        # meters
dx = 0.4       # error in distance
t = 0.4        # seconds
dt = 0.1       # error in time

# Average velocity
v = x / t

# Error propagation formula for division
dv = v * ((dx/x)**2 + (dt/t)**2)**0.5

print(f"Average velocity: {v:.2f} m/s")
print(f"Error in velocity: ±{dv:.2f} m/s")


Average velocity: 12.75 m/s
Error in velocity: ±3.34 m/s


## Exercise 2
[2] An enterprising cow attempts to jump over the moon by jumping vertically into the air with initial speed $v_0=4.0\pm0.2 m/s$. After a time $t=0.60\pm0.06s$, the height of the cow is $h = v_0t-\frac{1}{2}g t^2 = 0.636 m$. What is the uncertainty in $h$? Take $g$ as exactly $9.81 ms^{-2}$.

In [2]:
# given
v0, dv0 = 4.0, 0.2        # m/s
t, dt   = 0.60, 0.06      # s
g       = 9.81            # m/s^2 (exact)

# value of h
h = v0*t - 0.5*g*t**2

# uncertainty in h
dhdv0 = t
dhdt  = v0 - g*t
dh = ((dhdv0*dv0)**2 + (dhdt*dt)**2) ** 0.5

print(f"h = {h:.3f} m")
print(f"Δh = {dh:.2f} m")


h = 0.634 m
Δh = 0.16 m


## Exercise 3
[2] In an optics experiment the object distance $u$ is measured to be 20cm and the image distance $v$ is 10cm, both to an accuracy of 0.5cm. Find the focal length $f$ of the lens using the formula:

$$ \frac{1}{u}+\frac{1}{v}=\frac{1}{f}$$

In [4]:
# given values
u, du = 20, 0.5  # cm
v, dv = 10, 0.5  # cm

# focal length using lens formula
f = (u * v) / (u + v)

# partial derivatives
df_du = (v**2) / (u + v)**2
df_dv = (u**2) / (u + v)**2

# uncertainty in f
df = ((df_du * du)**2 + (df_dv * dv)**2)**0.5

print(f"Focal length f = {f:.2f} cm")
print(f"Uncertainty in f = ±{df:.2f} cm")



Focal length f = 6.67 cm
Uncertainty in f = ±0.23 cm


## Exercise 4
[2] Two students each measure the refractive index of water. Jack measures a value of $1.33 \pm 0.03$ while Jill measures $1.28 \pm 0.02$. Are these values in agreement? *You do have to think a bit about this one...*

In [6]:
# Given measurements
jack_value = 1.33
jack_error = 0.03

jill_value = 1.28
jill_error = 0.02

# Difference between values
difference = abs(jack_value - jill_value)

# Combined uncertainty
combined_error = (jack_error**2 + jill_error**2)**0.5

print(f"Difference = {difference:.3f}")
print(f"Combined uncertainty = {combined_error:.3f}")

if difference <= combined_error:
    print("The measurements agree within uncertainty.")
else:
    print("The measurements do not agree within uncertainty.")


Difference = 0.050
Combined uncertainty = 0.036
The measurements do not agree within uncertainty.


## Exercise 5
[2] The damped resonance frequency $\omega_{res}$ of an oscillating system is related to the (un-damped) natural angular frequency $\omega_0$ and the damping coefficient $\alpha$ by:

$$\omega_{res} = \sqrt{ \omega_0^2 - 2\alpha^2}$$

Find $f_0$ if the measured resonance frequency $f_{res}$ is $23.2\pm0.1 Hz$ and the measured damping coefficient is $19.5\pm0.5s^{-1}$.


In [7]:
import math

# given measurements
f_res, df_res = 23.2, 0.1      # Hz
alpha, d_alpha = 19.5, 0.5     # s^-1

# compute f0
g = 4*math.pi**2*f_res**2 + 2*alpha**2
f0 = (1/(2*math.pi))*math.sqrt(g)

# partial derivatives for uncertainty
df0_df = (2*math.pi*f_res)/math.sqrt(g)
df0_da = alpha/(math.pi*math.sqrt(g))

# uncertainty
df0 = math.sqrt( (df0_df*df_res)**2 + (df0_da*d_alpha)**2 )

print(f"f0 = {f0:.3f} Hz")
print(f"Δf0 = ±{df0:.3f} Hz")


f0 = 23.612 Hz
Δf0 = ±0.100 Hz


## Exercise 6
[2] Suppose you have the following equation from one of your lab experiments:

$$f=\frac{c}{2}\sqrt{\frac{n_x^2}{L_x^2}+\frac{n_y^2}{L_y^2}+\frac{n_z^2}{L_z^2}}$$

where $f$ is the resonant frequency of sound waves in a box of sides $L_x$, $L_y$ and $L_z$ in length and the $n_x$ etc. are integers. $L_x = 10.2\pm0.2m$, $L_y = 5.2\pm0.3m$ and $L_z = 20.0\pm0.1 m$, while $c = 331.3 + T * 0.606 \,m\,s^{-1}$ is the temperature-dependent speed of sound, and the temperature $T$ is $23 \pm 1^\circ C$.
Calculate $f$ and the error in $f$ for the following values of $(nx,ny,nz)= (1,1,1), (1,1,2)$ and $(2,1,1)$.


In [8]:
import math

# given values (m, s, °C)
Lx, dLx = 10.2, 0.2
Ly, dLy = 5.2, 0.3
Lz, dLz = 20.0, 0.1
T, dT = 23.0, 1.0

c  = 331.3 + 0.606*T          # m/s
dc = 0.606*dT                 # m/s

def f_and_df(nx, ny, nz):
    S = (nx**2)/(Lx**2) + (ny**2)/(Ly**2) + (nz**2)/(Lz**2)
    sqrtS = math.sqrt(S)
    f = (c/2)*sqrtS

    df_dc  = 0.5*sqrtS
    df_dLx = (c/2)*(-nx**2/(sqrtS*Lx**3))
    df_dLy = (c/2)*(-ny**2/(sqrtS*Ly**3))
    df_dLz = (c/2)*(-nz**2/(sqrtS*Lz**3))

    df = math.sqrt((df_dc*dc)**2 + (df_dLx*dLx)**2 +
                   (df_dLy*dLy)**2 + (df_dLz*dLz)**2)
    return f, df  # Hz

modes = [(1,1,1),(1,1,2),(2,1,1)]
for m in modes:
    f, df = f_and_df(*m)
    print(f"{m}: f = {f:.2f} ± {df:.2f} Hz")


(1, 1, 1): f = 38.25 ± 1.67 Hz
(1, 1, 2): f = 41.07 ± 1.56 Hz
(2, 1, 1): f = 48.19 ± 1.40 Hz


## Exercise 7
[4] The reflection coefficient $R_\parallel$ for parallel plane-polarised light reflected from a surface is given by the equation:

$$ R_\parallel = \frac{\tan^2(\theta_i - \theta_t)}{\tan^2(\theta_i + \theta_t)} $$

Calculate the error in $R_\parallel$ given measurements $\theta_i = (78 \pm 1)^\circ$ and $\theta_t = (40 \pm 1)^\circ$.

In [9]:
import math

# degrees -> radians
deg = math.pi/180

# inputs (degrees)
ti, dti = 78.0*deg, 1.0*deg
tt, dtt = 40.0*deg, 1.0*deg

A = ti - tt
B = ti + tt

# value of R_parallel
R = (math.tan(A)**2) / (math.tan(B)**2)

# helper for (sec^2 / tan)
def s2_over_tan(x):
    t = math.tan(x)
    return (1 + t*t)/t  # since sec^2 = 1 + tan^2

dlnR_dti = 2*(s2_over_tan(A) - s2_over_tan(B))
dlnR_dtt = 2*(-s2_over_tan(A) - s2_over_tan(B))

dR_dti = R * dlnR_dti
dR_dtt = R * dlnR_dtt

dR = math.sqrt((dR_dti*dti)**2 + (dR_dtt*dtt)**2)

print(f"R_parallel = {R:.3f}")
print(f"Uncertainty ΔR = ±{dR:.3f}")


R_parallel = 0.173
Uncertainty ΔR = ±0.027


## Exercise 8
[4] Calculate and print to the screen the fractional uncertainty, as a percentage to one
significant figure, of the fluid flow discharge coefficient $C_d$ from the equation

$$
C_d = \frac{\dot{m}\sqrt{1-\left(\frac{d}{D}\right)^4}}{Kd^2F\sqrt{\rho\Delta P}}
$$

where

\begin{align*}
    C_d &= \text{discharge coefficient}&& \text{(no units)} \\
    \dot{m} &= \text{mass flow rate}&& = 0.13 \pm 0.01kg\,s^{-1} \\
    d &= \text{orifice diameter}&& = 11\pm 1 mm \\
    D &= \text{pipe diameter}&& = 71 \pm 1 mm \\
    \rho &= \text{fluid density}&& =1.01\pm0.01g\,cm^{-3} \\
    \Delta P &= \text{differential pressure}&& =156 \pm 7 Pa \\
    K &= \text{a constant parameter}&& =\text{constant (no units)} \\
    F &= \text{thermal expansion factor}&& =\text{constant (no units)}
\end{align*}


In [10]:
import math

# inputs (values and 1-sigma uncertainties)
mdot, dmdot = 0.13, 0.01          # kg s^-1
d, dd       = 11e-3, 1e-3         # m
D, dD       = 71e-3, 1e-3         # m
rho, drho   = 1.01e3, 0.01e3      # kg m^-3  (1.01 ± 0.01 g/cm^3)
dP, ddP     = 156.0, 7.0          # Pa

# constants K and F are exact -> cancel in uncertainty
def Cd(mdot, d, D, rho, dP):
    return mdot*math.sqrt(1 - (d/D)**4)/(d**2 * math.sqrt(rho*dP))

# value
Cd_val = Cd(mdot, d, D, rho, dP)

# logarithmic derivatives for fractional-uncertainty propagation
Q = 1 - (d/D)**4
dlnC_dmdot = 1/mdot
dlnC_drho  = -0.5/rho
dlnC_ddP   = -0.5/dP
dQ_dd = -4*d**3/D**4
dQ_dD =  4*d**4/D**5
dlnC_dd = 0.5*(dQ_dd/Q) - 2/d
dlnC_dD = 0.5*(dQ_dD/Q)

# fractional uncertainty
frac = math.sqrt( (dlnC_dmdot*dmdot)**2 + (dlnC_drho*drho)**2 +
                  (dlnC_ddP*ddP)**2 + (dlnC_dd*dd)**2 + (dlnC_dD*dD)**2 )

print(f"C_d ≈ {Cd_val:.3f} (value, units cancel)")
print(f"Fractional uncertainty ≈ {100*frac:.1f}%")


C_d ≈ 2.706 (value, units cancel)
Fractional uncertainty ≈ 19.9%


## Exercise 9: Optional problem (not marked)
If you have time and want to try something interesting, do the following problem by plotting in 2D:
 - Draw an equilateral triangle with vertices and coordinates: vertex 1: $(p_1,q_1)$; vertex 2: $(p_2, q_2)$; vertex 3: $(p_3, q_3)$.
 - Place a dot at an arbitrary point $P = (x_0, y_0)$ within this triangle.
 - Find the next point by selecting randomly an integer $n = 1 , 2, $  or $3$ :
    1. If 1 , place a dot halfway between P and vertex 1.
    2. If 2 , place a dot halfway between P and vertex 2.
    3. If 3 , place a dot halfway between P and vertex 3.
 - Repeat the last two steps using the last dot as the new P.

Mathematically, the coordinates of successive points are given by the formulae

$$(x_{i+1},y_{i+1})=0.5[(x_i,y_i)+(p_n,q_n)]$$

and

$$n=int(1+3r_i),$$

where $r_i$ is a random number between 0 and 1 and where the $int()$ function outputs the closest integer smaller than or equal to the argument.

Try extending this to four vertices.